src/primitives/poseidon/p128pow5t3.rs

```diff
@@ -0,0 +1,199 @@
+use halo2::arithmetic::Field;
+use pasta_curves::pallas;
+
+use super::{
+    fp::{MDS, MDS_INV, ROUND_CONSTANTS},
+    Mds, Spec,
+};
+
+/// Poseidon-128 using the $x^5$ S-box, with a width of 3 field elements, and the
+/// standard number of rounds for 128-bit security "with margin".
+///
+/// The standard specification for this set of parameters uses $R_F = 8, R_P = 56$.
+/// This is conveniently an even number of partial rounds, making it easier to
+/// construct a Halo 2 circuit.
+#[derive(Debug)]
+pub struct P128Pow5T3;
+
+impl Spec<pallas::Base, 3, 2> for P128Pow5T3 {
+    fn full_rounds() -> usize {
+        8
+    }
+
+    fn partial_rounds() -> usize {
+        56
+    }
+
+    fn sbox(val: pallas::Base) -> pallas::Base {
+        val.pow_vartime(&[5])
+    }
+
+    fn secure_mds(&self) -> usize {
+        unimplemented!()
+    }
+
+    fn constants(
+        &self,
+    ) -> (
+        Vec<[pallas::Base; 3]>,
+        Mds<pallas::Base, 3>,
+        Mds<pallas::Base, 3>,
+    ) {
+        (ROUND_CONSTANTS[..].to_vec(), MDS, MDS_INV)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::marker::PhantomData;
+
+    use ff::PrimeField;
+    use halo2::arithmetic::FieldExt;
+    use pasta_curves::pallas;
+
+    use crate::primitives::poseidon::{permute, ConstantLength, Hash, Spec};
+
+    use super::{MDS, MDS_INV, ROUND_CONSTANTS};
+
+    /// The same Poseidon specification as poseidon::P128Pow5T3, but constructed
+    /// such that its constants will be generated at runtime.
+    #[derive(Debug)]
+    pub struct P128Pow5T3<F: FieldExt> {
```

Comment:
Make it clear that this it runtime-generated.

```suggestion
    pub struct P128Pow5T3Gen<F: FieldExt> {
```